Patch: Fix Index Resetting on Removal, Fix bad `isShuffled` reference

johnnyshankman opened the pull request:
## This PR

Fixes a bug causing the index/trackNumber of the currently playing song to be incorrect after calling `removeTrack`.

Found while developing on my music app *hihat*:
* https://github.com/johnnyshankman/hihat/pull/41

## Scope

- **`removeTrack(index)` no longer resets `trackNumber` to 0** when removing any track in the tail half of the list. Root cause was an over-decrement in `Gapless5FileList.remove` whenever `index >= numTracks - 2`, regardless of whether the removed track was before or after the current one.
- **Removing the currently-playing track** now correctly auto-advances to the track that slid into the removed slot (or clamps to the new last track if the current was the tail, or sets `trackNumber` to `-1` if the list is now empty). Playback resumes on the new current if it was playing before. `onnext` fires so UI listeners can repaint.
- **Removing the currently-playing track repeatedly until empty** no longer crashes in `updateLoading()` on an undefined source. `updateLoading()` is now skipped when the list is empty.
- **`shuffledIndices` is preserved** across removes (prior iteration of this PR wiped it — fixed per review feedback).

## Reproduction of the original issue

1. Load a track list of three songs.
2. `getIndex()` returns `0`.
3. `gotoTrack(1)` — `getIndex()` returns `1`.
4. `removeTrack(2)` — removes the tail.
5. Before this PR, `getIndex()` returned `0` instead of `1`, causing the player to pause and return wrong track info.

## Note on `replaceTrack`

`replaceTrack` defers to `removeTrack` + `insertTrack`, and this PR does not attempt to fix `replaceTrack` behavior when replacing the currently-playing track — the removeTrack auto-advance + insertTrack re-shift can leave `trackNumber` pointing at the wrong source in that specific case. That's a pre-existing interaction issue worth a separate follow-up PR.

## QA

- New regression tests cover:
  - Removing a track after the currently playing track preserves the index.
  - Removing the currently playing track auto-advances (including walking down to empty).
  - Removing a track keeps `shuffledIndices` intact.
  - Removing the current track repeatedly until empty does not crash.
- All existing tests still pass.

<img width="878" height="389" alt="Screenshot 2026-04-13 at 6 14 01 PM" src="https://github.com/user-attachments/assets/b106adad-95f2-45f1-b570-9e92642c1503" />

johnnyshankman commented on types/gapless5.d.ts:
```diff
@@ -331,7 +331,7 @@ declare class Gapless5Source {
     getPosition: () => number;
     getLength: () => number;
-    play: (syncPosition: any) => void;
+    play: (syncPosition: any, skipCallback: any) => void;
```
Note: I reran the type generator and this just appeared.

johnnyshankman commented:
Tests aren't passing. Not entirely sure why...

johnnyshankman commented:
Unsure why this one seemingly random test has scope issues now:
```

  ● Gapless-5 object with tracklist › triggers navigation callbacks

    ReferenceError: state is not defined

      1613 |
      1614 |   this.stop = (resetPosition = false) => {
    > 1615 |     if (state === Gapless5State.None) {
           |     ^
      1616 |       return;
      1617 |     }
      1618 |     log.debug(`Stopping: ${this.audioPath}`);

      at Gapless5.stop (gapless5.js:1615:5)
      at Object.<anonymous> (gapless5.test.js:152:12)
      ```

johnnyshankman commented:
Any feedback @regosen would love to get this into main! Happy to do whatever you need. This has conflicts as it is old now but I can update it anytime with the latest package and types.

Been using it for some time with hihat and it's been lovely. I'm going to actually add one more feature to allow for swapping audio outputs. Thank you for the open-source code either way!

regosen commented on gapless5.js:
```diff
@@ -1331,36 +1331,39 @@ function Gapless5(options = {}, deprecated = {}) { // eslint-disable-line no-unu
-      curSource.stop();
+    // If it's the current track, pause playback
+    if (wasPlayingCurrentTrack) {
+      this.pause();
```
I'm not sure this is what all users would want the player to do.  I can imagine wanting to jump to the next track and continue playing.  How does it behave without this PR if you delete the currently playing track?

johnnyshankman replied:
Good call, let me try to maintain the original UI/UX contract here as best I can.

Been a minute, but if I remember correctly I was noticing behavior where audio would pause indefinitely, along with getting back incorrect track info (OBO issue). 

So for my own knowledge, the UI contract I'm going to respect reads like:
  - On remove-current: move to the track that now occupies the same index (or wrap/clamp if we removed the last).
  - If it was playing before removal: keep playing the new occupant.
  - Only pause if the list becomes empty.

regosen commented on gapless5.js:
```diff
@@ -1331,36 +1331,39 @@ function Gapless5(options = {}, deprecated = {}) { // eslint-disable-line no-unu
-      log.warn(`Cannot remove missing track: ${pointOrPath}`);
+    const index = typeof pointOrPath === 'number' ? pointOrPath : this.findTrack(pointOrPath);
+    if (!isValidIndex(index)) {
+      log.error(`Cannot remove missing track at index: ${index}`);
```
This is inconsistent with other log messages concerning missing tracks.  In other projects they are referenced by path and I've used the debug console to look for the path instead of the index.  What was the advantage of changing it?

johnnyshankman replied:
👍🏼  reverting

regosen commented on gapless5.js:
```diff
@@ -1331,36 +1331,39 @@ function Gapless5(options = {}, deprecated = {}) { // eslint-disable-line no-unu
 
+    this.playlist.sources[index].unload();
+    this.playlist.sources.splice(index, 1);
+    this.playlist.shuffledIndices = [];
```
Why is it desired to empty out `shuffledIndices` here?  I would expect that to break things in shuffle mode.

johnnyshankman replied:
Ah yep this does break shuffle as I'm nuking the shuffle order on every remove. Fixing!

regosen commented:
Thanks @johnnyshankman for following up and finding a bug!  I just left some comments.  Hihat looks really cool!

johnnyshankman commented:
Will go over your comments tomorrow and see what changes need to be made! Thanks for reviewing.

Update: Great review, you're right on all 3 concerns. Made a really naive solution without shuffle etc in mind. Working on a new proposal!

johnnyshankman committed:
Fix removeTrack index-reset bug and handle current-track removal

Rework Gapless5FileList.remove and Gapless5.removeTrack to fix three
related issues:

1. Removing any track in the tail half of the list would reset the
   current track index to 0.  Root cause was in Gapless5FileList.remove,
   which over-decremented trackNumber whenever
   `index >= numTracks - 2`, regardless of whether the removed track was
   before or after the current one.

2. Removing the currently-playing track would leave trackNumber pointing
   at a stale or invalid slot.  The remove path now either auto-advances
   to the track that slid into the removed slot, clamps to the new last
   track (when the current was the tail), or sets trackNumber to -1
   (when the list is empty).

3. Calling updateLoading() after emptying the list would crash with a
   TypeError when dereferencing an undefined source.  Now guarded.

removeTrack continues to resume playback on the new current track when
the removed track was actively playing, matching the pre-existing
auto-advance UX.

Tests:
- Regression: removing a track after the currently playing track keeps
  the correct index.
- Regression: removing the currently playing track auto-advances.
- Regression: removing a track preserves shuffledIndices.
- Regression: removing current track repeatedly until empty does not
  crash.

johnnyshankman commented:
Rebased and force pushed to fix conflicts.

Sidenote: Since we have unit tests and linting, we should run them on all PRs using github action, and make them block merge if they are not passing. For now, here's the screen proving no regressions in existing test cases and new regression tests area passing.
<img width="878" height="389" alt="Screenshot 2026-04-13 at 6 14 01 PM" src="https://github.com/user-attachments/assets/b106adad-95f2-45f1-b570-9e92642c1503" />

Did some manual playwright tests as well

<img width="1039" height="411" alt="Screenshot 2026-04-13 at 6 39 10 PM" src="https://github.com/user-attachments/assets/7342c576-d0c1-4d9d-8e63-4fe9d6bf5f6d" />

johnnyshankman committed:
Fire onnext when removeTrack auto-advances current track

Capture deleted source's audioPath before removal so the onnext
callback can fire with (deletedPath, newPath) when removeTrack
causes the current track to silently swap to the slid-in neighbor.
Previously the direct play() call bypassed onnext, leaving UI
listeners showing stale track metadata.

Also trim trailing blank lines in gapless5.test.js.

johnnyshankman committed:
Bump version to 1.6.2

regosen commented:
Oh wow, @johnnyshankman you really rewrote the entire change!  I agree this looks much better now.  I just replaced the library in my demo page + static projects and I'm testing it out end-to-end.

That's a good idea regarding github actions.  In the meantime, would you mind fixing the linter errors on `gapless5.test.js`?  Of particular concern are the ones with `'player' is already declared in the upper scope`.

regosen commented:
One more thing: I have a looped playlist and I just tried removing the final track while it was playing.  It switched to playing the previous track instead of the "next" one (which in my case would've been the first track since it was looped).  Is that by design?

johnnyshankman commented:
> One more thing: I have a looped playlist and I just tried removing the final track while it was playing. It switched to playing the previous track instead of the "next" one (which in my case would've been the first track since it was looped). Is that by design?

Correct! That is in my scenario 3 from e2e testing, that it clamps down to the new end of the list. That being said, open to adjusting the behavior when loop is engaged if you think instead of clamping on the tail, we wrap back around to 0.

regosen commented:
Yes, I think it would be more consistent if the player wrapped back around if loop is engaged.  If loop is not engaged when removing the last track it should just stop.  This is also consistent with what Apple Music does (well sort of- Apple Music continues playing the removed track, though I don't expect to do that here).

johnnyshankman commented:
@regosen awesome, will get that happening ASAP. makes perfect sense to me. removing current when it's the tail and loop is engaged (but not single mode) we should wrap. when loop is not engaged, we should clamp *and pause*.

for single mode... i think the same behavior makes sense?

johnnyshankman committed:
Tail-remove: wrap to 0 when loop on, pause when loop off

When removeTrack removes the currently-playing track and it was at
the tail of the list:
  - loop on:  wrap trackNumber back to 0 and resume playback on the
              new first track. Emits onnext with (deletedPath, firstPath).
  - loop off: clamp trackNumber to the new last track (unchanged) but
              always pause instead of auto-resuming, even if playback
              was active before the removal.

singleMode has no bearing on this logic — behavior depends purely on
loop. This keeps the contract symmetric with next() at the end of the
playlist.

Tests:
- Regression: removing current-at-tail with loop=true wraps to start
  and resumes (asserts onplayrequest fires on the wrapped track).
- Regression: removing current-at-tail with loop=false clamps and
  always pauses (asserts onplayrequest is NOT called).

regosen commented:
> for single mode... i think the same behavior makes sense?

Yes @johnnyshankman , I think this makes sense.  I'm trying this out now and I just noticed another bug (not a regression though): if I delete a previous track while playing, it often stops.  (e.g. deleting track at index 0 while playing index 1).  I think the issue is that `this.play();` is inside the `if (deletedPlaying...` check.  But we should resume playing regardless.

Also, `yarn lint` still shows a bunch of warnings on gapless5.test.js, can you fix those?

johnnyshankman committed:
Fix all lint warnings in gapless5.test.js

- Add spaces inside array brackets (array-bracket-spacing)
- Remove blank line padding before closing brace (padded-blocks)
- Rename inner player vars to testPlayer to avoid shadowing (no-shadow)

johnnyshankman commented:
Good call @regosen my bad! Forgot to lint fix. Should be good now.

And on the note with that bug, that's a good find. We should tackle that next, I'll take a crack in a future PR if ya don't beat me there 😛

regosen commented:
In that case I've opened a new issue for that bug, and it's worse than I thought- it also stops if I delete a future track: https://github.com/regosen/Gapless-5/issues/63

That being said, I think I found a regression: if I remove a track even if stopped then on the subsequent play it stops mid-way through the next track.  This seems to be intermittent, and I'm not experiencing this on main.  Are you able to reproduce this?  I'm using a modified version of my Gapless5 demo page where I call `removeTrack(0)` before playing.

johnnyshankman commented:
will give it a test run tonight/tomorrow and see what i find. i'm amassing a little QA list to potentially automate down the line. thanks for the thorough testing!

regosen commented:
Sounds good, thanks!  FWIW I was testing with a 3-track playlist.  When it goes from 3 to 2 tracks I get a console line about shuffle mode being disabled (even if it was already disabled).  Could be related.

johnnyshankman committed:
Fix isShuffled() call in FileList.remove

`this.isShuffled` is a function reference on the FileList — always
truthy — so `if (this.isShuffled && !player.canShuffle())` was
tripping on every removal that dropped the count below the shuffle
minimum, invoking `setShuffle(false)` (and flipping uiDirty) even
when shuffle had never been engaged. Call the function: `isShuffled()`.

Pre-existing bug on main; opportunistic fix.

Test: asserts playlist.setShuffle is not called on a non-shuffled
player that drops from 3 tracks to 2.

johnnyshankman commented:
@regosen found the shuffle bug and snuck the patch in!

both main and my branch had the same bug. a conditional statement checking for `this.isShuffled` which is a function reference and therefore always evals to true. easy fix, swapped to `this.isShuffled()` to get the actual boolean we want

i can't yet recreate your bug with the skip in the middle of a currently playing song. can you retest and also send me some super clear repro instructions? would love to squash if i can prove i've squashed it.

my only hunch would be maybe it has something to do with the new call to `this.onnext` when we remove but i'm not sure why it would be so delayed like in your bug description. maybe it's to do with cross fade.

regosen reviewed (approved):
Confirmed no regressions!
